docs/question/server.md

```diff
@@ -1,17 +1,104 @@
+
+### Step 2: `prepare`
+
+Next, the `prepare` function is called after all elements (e.g. `<pl-number-input>`) have run `generate()`. This is typically done to do any sort of final post-processing, but is not commonly used.
+
+### Step 3: `render`
+
+Next, the `render(data, html)` function is called to render the question. You can use this function to override how the question is rendered. The render function expects two parameters, `data` and `html`, and should return a string of HTML. The HTML after the mustache template has been rendered is available through the `html` parameter. This is typically only used for more advanced questions, and we won't need it for this example.
+
+### Step 4: `parse`
+
+When a student submits their answer, the `parse` function is called to parse the submitted answers after the individual elements have parsed them. This function can be used to display more-specific format errors than the input elements or to parse the input differently. When our parse function runs, `<pl-number-input>` will have already parsed the submitted value as an integer, and display an error to the student if it was invalid. For this question, we will allow the student to only submit positive integers, so we will check for that and set a format error with `data["format_errors"]` if it is negative.
+
+- `data["raw_submitted_answers"][NAME]` is the exact raw answer submitted by the student.
+- `data["submitted_answers"][NAME]` is the answer parsed by elements (e.g. strings converted to numbers).
+- `data["format_errors"][NAME]` is the format error message for a named answer.
+
+If there are format errors, then the submission is "invalid" and is not graded. To provide feedback but keep the submission "valid", `data["feedback"][NAME]` can be used instead.
+
+```python title="server.py"
+def parse(data):
+    # Reject negative numbers for "y" if we don't already have a format error
+    if "y" not in data["format_errors"] and data["submitted_answers"]["y"] < 0:
+        data["format_errors"]["y"] = "Negative numbers are not allowed"
+```
+
+### Step 5: `grade`
+
+Finally, the `grade(data)` function is called to grade the question. The grade function is responsible for:
+
+- Setting the score and feedback for each named answer in `data["partial_scores"]`.
+- Setting the total score for the question in `data["score"]`.
+- Setting the overall feedback for the question in `data["feedback"]`.
+
+This function only runs if `parse()` did not produce format errors, so we can assume all data is valid. All elements will have already graded their answers (if any) before this point.
+
+- `data["partial_scores"][NAME]["score"]` is the individual scores for the named answers (0 to 1).
```

Comment:
I think you should also point out `data["partial_scores"][NAME]["feedback"]`, which is the element-specific feedback (as opposed to the question-wide feedback).

Reply:
Pointed this out in 189fcab7d634a129d31bfb5c5fd1de4e2ee3cbd9. I also added a note that instructors shouldn't use this field for giving feedback.

Reply:
I don't know if I agree with that note. We use the partial scores feedback in our own example course questions.

https://github.com/PrairieLearn/PrairieLearn/blob/b1929d0341d01e6ed1d94bfb3e1293a89a435d9a/exampleCourse/questions/demo/annotated/LectureVelocity/4-KnowledgeTest/server.py#L90-L102

One issue with `data["feedback"][NAME]` is that it is not shown by default, and instructors need to explicitly show it using `{{feedback.NAME}}`. The partial scores feedback can be shown by specific elements in an element-specific way.

https://github.com/PrairieLearn/PrairieLearn/blob/b1929d0341d01e6ed1d94bfb3e1293a89a435d9a/apps/prairielearn/elements/pl-number-input/pl-number-input.mustache#L132-L136

Reply:
Ah I wasn't aware of that. That is a very important detail!

Reply:
Updated in 21cca4540216bd25e2fa512cf2f8f4a3e2f34a7f.

Reply:
@jonatanschroeder I updated the feedback section further in f1cc70d4a6303cb66a9ad081fdc705ef1daa184e.